text/0000-intra-rustdoc-links.md

```diff
@@ -0,0 +1,389 @@
+i.e., it has to start with an URI scheme.
+Thus, it will not be possible to write `<Foo>`
+to link to a Rust item called `Foo`
+that is in scope
+(this also conflicts with Markdown ability to contain arbitrary HTML elements).
+And while `<std::iter::Iterator>` is a valid URI
+(treating `std:` as the scheme),
+to avoid confusion, the RFC does not propose adding any support for autolinks.
+
+[absolute URI]: http://spec.commonmark.org/0.27/#absolute-uri
+
+This means that this will not render a valid link:
+
+```
+Does not work: <bars::Bar> :(
+```
+
+We suggest to use [Implied Shortcut Reference Links][isrl] instead:
+
+```
+Does work: [`bars::Bar`] :)
+```
+
+which will be rendered as
+
+```html
+Does work: <a href="../bars/struct.Bar.html"><code>bars::Bar</code></a> :)
+```
+
+## Resolving paths
+
+The Rust paths used in links are resolved
+relative to the item in whose documentation they appear.
+
+For example:
+
+```rust
+/// Container for a [Dolor](ipsum::Dolor).
+struct Lorem(ipsum::Dolor);
+
+/// Contains various things, mostly [Dolor](ipsum::Dolor) and a helper function,
+/// [sit](ipsum::sit).
+mod ipsum {
+    struct Dolor;
+
+    /// Takes a [Dolor] and does things.
+    fn sit(d: Dolor) {}
+}
+
+mod amet {
+  //! Helper types, can be used with the [ipsum](super::ipsum) module.
+}
+```
+
+## Path ambiguities
+[path-ambiguities]: #path-ambiguities
+
+Rust allows items to have the same name. A short evaluation revealed that
+
+- unit and tuple struct names
+  can conflict with
+  function names,
+- unit and tuple struct names
+  can conflict with
+  enum names,
+- and regular struct, enum, and trait names
+  can conflict with
+  each other
+  (but not function names).
+
+It appears the ambiguity can resolved
+by being able to restrict the path to function names.
+We propose that in case of ambiguity,
+you can add `()` as a suffix to path
+to only search for functions.
+(Additionally, to link to macros, you must add `!` to the path.)
+
+This was originally proposed by
+[@kennytm](https://github.com/kennytm)
+[here](https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284719684),
+going into more details:
+
+> `<syntax::ptr::P>` — First search for type-like items. If not found, search for value-like items
+>
+> `<syntax::ptr::P()>` — Only search for functions.
+>
+> `<std::stringify!>` — Only search for macros.
+
+## Linking to external crates
+
+Rustdoc is already able to link to external crates,
+and renders documentation for all dependencies by default.
+Referencing the standard library (or `core`)
+should generate links with a well-known base path,
+e.g. `https://doc.rust-lang.org/nightly/`.
+Referencing other external crates
+links to the pages Rustdoc has already rendered (or will render) for them.
+Special flags (e.g. `cargo doc --no-deps`) will not change this behavior.
+
+## Errors
+
+Ideally, Rustdoc would be able to recognize Rust path syntax,
+and if the path cannot be resolved,
+print a warning (or an error).
```

Comment:
Docs generate urls, wether they're valid or not. rustdoc won't generate docs if the code doesn't compile (so warnings are fine).